### PR TITLE
Pd 13298 popover focus

### PR DESCRIPTION
**Pull request by jamiek-acl**
Bug fix in the Popover component.

### 🛠 Purpose

Before this PR, when you open the popover, focus goes into it.  Then when you tab through its content, focus goes to the address bar (since the Popover is added to the bottom of the DOM).

After this PR, when you open the popover, focus goes into it. Then when you tab through its content, focus goes to the next item on screen and the popover closes.

This is maybe premature and it needs to be tested with `isEager` and some other props.... just want to get some feedback.

### 🖥 Screenshots

Notice the `tabIndex` before (3), how it is removed (-1), then how it is restored (3).
Notice I can tab right through the popover, in both directions.

focus2

### CODE REVIEW

File: packages/Popover/src/Popover.js [hunk 1]
```diff
@@ -26,6 +26,10 @@ import PopoverStyled from "./Popover.styles";
 const openDelay = 350;
 const closeDelay = 150;
 const throttleDelay = 20;
+const focusableElementSelector = 'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';
```
Comment (oscarkwan) [inline packages/Popover/src/Popover.js]: Will our Buttons that are (spans) work?

Comment (jamiek-acl) [reply packages/Popover/src/Popover.js]: They'd have to have `tabIndex` so ya

Comment (nahumzs) [reply packages/Popover/src/Popover.js]: do we need to add `[role="button"]`?

Comment (jamiek-acl) [reply packages/Popover/src/Popover.js]: Makes sense to me

File: packages/Popover/src/Popover.js [hunk 2]
```diff
@@ -175,6 +180,8 @@ class Popover extends React.Component {
     // about setState for Popovers and Tooltips in ComponentDidMount
     // https://reactjs.org/docs/react-component.html#componentdidmount
 
+    focusableElements = document.querySelectorAll(focusableElementSelector);
```
Comment (mikrotron) [inline packages/Popover/src/Popover.js]: I think this should be set when the Popover opens.  The elements on the page could vary while the Popover is mounted.

File: packages/Popover/src/Popover.js [hunk 3]
```diff
@@ -26,6 +26,11 @@ import PopoverStyled from "./Popover.styles";
 const openDelay = 350;
 const closeDelay = 150;
 const throttleDelay = 20;
+const focusableElementSelector =
+  'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"]), [role="button"]';
+let focusableElements;
+const originalTabIndexes = [];
+let triggerFocusIndex = -1;
```
Comment (mikrotron) [inline packages/Popover/src/Popover.js]: Why `-1`? Can it just be left undefined?

**Comment by jamiek-acl** [comment]
Aside: I just noticed that if `isEager` or `shouldKeepFocus` is true, you can't get into the content of the Popover with the tab keys (e.g. if it had a link in it). This is true for the existing Popover, and also for this revised one. Is that a problem?

Also, what's the difference between `isEager` and `shouldKeepFocus`...? They seem to have the same purpose, and are only checked together [in the code].

**Comment by mikrotron** [comment]
@jamiek-acl I think `shouldKeepFocus` still requires a click to open, it just doesn't transfer the focus to the `<Content>`.  This is important for the `<DatePicker>` I think, but @allison-c knows better than me.

I think it's fine that `isEager`, tooltip-style `<Popovers>` don't allow focus to their contents.  The content is linked to the trigger element by `aria-describedby`, and our design patterns do not support having interactive elements in a hover-activated tooltip (I think). One of the designers would be able to verify that though.

File: packages/Popover/src/Popover.js [hunk 4]
```diff
@@ -18,14 +16,11 @@ import Tip from "./components/Tip/Tip";
...
-// ACCESSIBILITY
-// NOTE: When closing the popover seems to be better to focus the trigger button
-//       only when the close method is the ESC key, when clicking outside doesn't feel a fluid action
-//       assigning the focus to the button again
-
 const openDelay = 350;
 const closeDelay = 150;
 const throttleDelay = 20;
+const focusableElementSelector =
+  'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"]), [role="button"]';
```
Comment (mikrotron) [inline packages/Popover/src/Popover.js]: Just `'button, [href], input, select, textarea, [tabindex]'` should work, and is "cleaner", but 🤷‍♂️

Comment (jamiek-acl) [reply packages/Popover/src/Popover.js]: Yeah I saw that one. It is a bit easier to read, but no point in touching elements that already have the value we want

File: packages/Popover/src/Popover.js [hunk 5]
```diff
@@ -181,8 +180,15 @@ class Popover extends React.Component {
     }
   }
 
+  componentWillReceiveProps(nextProps) {
```
Comment (alexzherdev) [inline packages/Popover/src/Popover.js]: This lifecycle is deprecated. Perhaps in the future we can migrate this to `getDerivedStateFromProps`.

File: packages/Popover/src/Popover.js [hunk 6]
```diff
@@ -296,11 +302,66 @@ class Popover extends React.Component {
...
+      }
+      elementsParent = elementsParent.parentNode;
+    }
+    return false;
+  };
+
+  restoreTabIndexes = () => {
+    this.focusableElements.forEach(focusableElement => {
+      if (focusableElement !== this.$trigger && !this.elementIsDescendentOfPopover(focusableElement)) {
+        focusableElement.tabIndex = this.originalTabIndexes.reverse().pop(); // eslint-disable-line no-param-reassign
```
Comment (alexzherdev) [inline packages/Popover/src/Popover.js]: `reverse()` is in-place, and this is done in a loop, so probably there's an issue with reversing multiple times. `this.originalTabIndexes.shift()` should work?

**Comment by nahumzs** [comment]
`shouldKeepFocus` works when you want to have as example a search, where the input will open the popover but will keep the focus on it. 
DatePicker, ListBoxWithTags, ComboBox are some of the possible components that can and will benefit from having that prop. @jamiek-acl

File: packages/Popover/src/Popover.js [hunk 7]
```diff
@@ -296,11 +308,64 @@ class Popover extends React.Component {
...
+    });
+  };
+
+  removeTabIndexes = () => {
+    this.focusableElements.forEach((focusableElement, index) => {
+      if (focusableElement === this.$trigger) {
+        this.triggerFocusIndex = index;
+      } else if (!this.elementIsDescendentOfPopover(focusableElement)) {
+        this.originalTabIndexes.push(focusableElement.tabIndex);
+        focusableElement.tabIndex = -1; // eslint-disable-line no-param-reassign
```
Comment (mikrotron) [inline packages/Popover/src/Popover.js]: Did you try out removing the `tabindex` manipulation? This is the part that makes me the most nervous (this individual component alters a lot of external DOM elements, which isn't very component-like).

Comment (jamiek-acl) [reply packages/Popover/src/Popover.js]: Latest push does not touch the other items' tabIndex

**Pull request merged**